### ISSUE TITLE
Login form doesn't use labels

### ISSUE THREAD

**Issue by gnomeontherun** [vmware/clarity#3522]
## Describe the bug

Login form fields do not have labels, and screen readers will not understand their roles.

## How to reproduce

https://clarity.design/documentation/login

- Using a screen reader, read any of the fields in the 'Login Page' example and observe that they are not labelled.

## Expected behavior

Use <label> elements with the for attribute to provide visible textual labels for all input fields.

## Versions

**App**

* Angular: [e.g. 6]
* Node: [e.g. 8.10.0]
* Clarity: [e.g. 0.12.5]

**Device:**

* Type: [e.g. MacBook]
* OS: [e.g. iOS]
* Browser [e.g. chrome, safari]
* Version [e.g. 22]

## Additional notes

This changes the login layout and should have UX input before implementation.

**Comment by colinreedmiller** [comment]
Visual design issue related to the original design: Underlined form fields feature a left indent and labeled form fields in login require a vertical arrangement (label above input). This results in a jagged left justification which has been judged undesirable. This is the reason the current (unlabeled) design is as it is. Adding labels without any visual design change (left padding on underlined input) with reveal this issue.

Possible solution: remove left padding for this case.

**Comment by mathisscott** [comment]
Possible solution 2? Minimize left padding? The current left padding is 6px. If 0px looks odd, maybe a 1 or 2px value would work?

**Comment by colinreedmiller** [comment]
Good idea with the reduction to 2px - a quick exploration suggests the offset from the label is enough that the zig-zag continues to be apparent. Another possibility is a change in the vertical rhythm that de-emphasizes the back and forth effect by structuring the design such that each label / input pair reads as a cohesive thought, with sufficient padding between it and the next.

**Comment by mathisscott** [comment]
A change to the vertical rhythm/baseline grid may be out of scope for this though?

**Comment by colinreedmiller** [comment]
@mathisscott - The initial guidance for using the left padding on the input was to create a visual cue that differentiates label from input. Research suggests that it is common to not consider this an issue and instead enforce an even left alignment.

IMO we have a visual design issue where either:
A. An problem is perceived where one is not present - the user clearly recognizes the distinction of label vs input without the left padding.
B. Another aspect of the design is leading to the problem and the left padding functions as a band-aid and leads to other problems such as this ticket.

re: Changing the vertical rhythm. This may be out of scope but may also be wise to consider given the stand-alone aspect of the login component and the fact that we are in there already. The issues about consistency between products, perceived security (looks different - is this authentic?), vertical space, etc are important and may require enough adjustment that the vertical rhythm question becomes moot.

I suggest we look at a more complete login analysis since we also have an open issue regarding the visual design for the right half of the screen.

**Comment by clane** [comment]
When looking at the login page documentation page, I notice there are now labels present but they are visually hidden with .clr-sr-only. Can we re-open this issue?

**Comment by colinreedmiller** [comment]
@clane If I remember correctly - the solution of adding sr-only labels was chosen as the most direct path to compliance without the additional effort required to redesign the login. Crucial here is that the login functions as a trust portal for users, and for this reason the impact of visual changes needs to be carefully assessed before being undertaken.

**Comment by sheribyrne** [comment]
@colinreedmiller: WCAG requires that "sufficient cues to determine the purpose" be present to get away with invisible labels.  They further describe that as "context and visual appearance of the control make its purpose clear"  Is that the case here?  (sorry coming into this mid-stream)

**Comment by colinreedmiller** [comment]
@sheribyrne - In the most recent team discussion on this issue, the understanding that the decision to include hidden labels was compliant. Looking at the documentation, the hint text remains when the input is focused, only disappearing when the user begins typing - perhaps that complies with "context and visual appearance of the control make its purpose clear"

**Comment by gnomeontherun** [comment]
Additionally, there was strong resistance to changing the visual behavior of the login screen at this time. The report was focused on screen reader behaviors which are resolved by having hidden labels, so I believe we should create a different ticket with updated specifics if there continues to be a concern.

**Comment by clane** [comment]
Taking another look at the Login Page and found that in Firefox on Windows, the contrast ratio of the placeholder is low. The foreground color I found when testing with the [Colour Contrast analyser](https://developer.paciellogroup.com/resources/contrastanalyser/).  It shows that the foreground color is #74A5D1 and the background color is #FAFAFA. These colors yield a contrast ratio of 2.5:1 which is far below the required ratio of 4.5:1.  This prevents us from using placeholder attributes as visual labels because low vision users will not be able to read these labels.

**Comment by chlane** [comment]
It looks like WCAG does require visible labels after all as described in [Success Criterion 3.3.2 (Level A)](https://www.w3.org/TR/WCAG21/#labels-or-instructions) Labels or instructions. See - https://www.w3.org/WAI/WCAG21/Techniques/html/H44.html and the related issue https://github.com/vmware/clarity/issues/4639

**Comment by mathisscott** [comment]
@chlane 
This would require a re-design of that screen which would then involve some sort of roll-out to product teams. It's not a simple fix of adding labels where there are none. It would also constitute a visual breaking change – so it could not be backported and would have to be released in a major version. Additionally, the earliest it could go in is 5.0. We don't have enough time left in the 4.0 cycle.

**Comment by gnomeontherun** [comment]
Clarity cares deeply about accessibility, and we work with our accessibility team to identify and triage issues. We have an internal tracking system to identify and verify accessibility issues, which has captured this already. Some issues with Clarity Angular are difficult to solve without causing side effects or breaking changes, and we have to balance concerns with fixing issues and breaking changes. As we build out Clarity Core, we are able to address these root cause accessibility issues and recommend adopting the Clarity Core implementations as they become available. We are closing this issue as we have the issues tracked internally. Please follow our development and releases to see when we release relevant Clarity Core components to make this possible.
